src/kem.c

```diff
@@ -0,0 +1,393 @@
+struct p11prov_kem_ctx {
+    P11PROV_CTX *provctx;
+    P11PROV_OBJ *key;
+    int op;
+};
+typedef struct p11prov_kem_ctx P11PROV_KEM_CTX;
+
+DISPATCH_MLKEM_FN(newctx);
+DISPATCH_MLKEM_FN(freectx);
+DISPATCH_MLKEM_FN(encapsulate_init);
+DISPATCH_MLKEM_FN(encapsulate);
+DISPATCH_MLKEM_FN(decapsulate_init);
+DISPATCH_MLKEM_FN(decapsulate);
+DISPATCH_MLKEM_FN(set_ctx_params);
+DISPATCH_MLKEM_FN(settable_ctx_params);
+static int p11prov_mlkem_init(void *vctx, int op, void *vkey,
+                              const OSSL_PARAM params[]);
+
+static void *p11prov_mlkem_newctx(void *provctx)
+{
+    P11PROV_CTX *ctx = (P11PROV_CTX *)provctx;
+    P11PROV_KEM_CTX *kemctx;
+
+    kemctx = OPENSSL_zalloc(sizeof(P11PROV_KEM_CTX));
+    if (kemctx == NULL) {
+        return NULL;
+    }
+
+    kemctx->provctx = ctx;
+    return kemctx;
+}
+
+static void p11prov_mlkem_freectx(void *vctx)
+{
+    P11PROV_KEM_CTX *ctx = vctx;
+
+    if (ctx == NULL) {
+        return;
+    }
+
+    p11prov_obj_free(ctx->key);
+    OPENSSL_clear_free(ctx, sizeof(P11PROV_KEM_CTX));
+}
+
+static int p11prov_mlkem_init(void *vctx, int op, void *vkey,
+                              const OSSL_PARAM params[])
+{
+    P11PROV_KEM_CTX *ctx = vctx;
+    P11PROV_OBJ *key = vkey;
+    CK_RV ret;
+
+    if (vctx == NULL || vkey == NULL) {
+        return RET_OSSL_ERR;
+    }
+
+    ret = p11prov_ctx_status(ctx->provctx);
+    if (ret != CKR_OK) {
+        return RET_OSSL_ERR;
+    }
+
+    p11prov_obj_free(ctx->key);
+    ctx->key = p11prov_obj_ref(key);
+    if (ctx->key == NULL) {
+        P11PROV_raise(ctx->provctx, CKR_ARGUMENTS_BAD, "Invalid object");
+        return RET_OSSL_ERR;
+    }
+
+    if (p11prov_obj_get_key_type(ctx->key) != CKK_ML_KEM) {
+        P11PROV_raise(ctx->provctx, CKR_KEY_TYPE_INCONSISTENT,
+                      "Not an ML-KEM key");
+        return RET_OSSL_ERR;
+    }
+
+    ctx->op = op;
+    return p11prov_mlkem_set_ctx_params(vctx, params);
+}
+
+static int p11prov_mlkem_encapsulate_init(void *vctx, void *vkey,
+                                          const OSSL_PARAM params[])
+{
+    P11PROV_OBJ *key = vkey;
+
+    if (p11prov_obj_get_class(key) != CKO_PUBLIC_KEY) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return RET_OSSL_ERR;
+    }
+    return p11prov_mlkem_init(vctx, EVP_PKEY_OP_ENCAPSULATE, key, params);
+}
+
+static int p11prov_mlkem_decapsulate_init(void *vctx, void *vkey,
+                                          const OSSL_PARAM params[])
+{
+    P11PROV_OBJ *key = vkey;
+
+    if (p11prov_obj_get_class(key) != CKO_PRIVATE_KEY) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return RET_OSSL_ERR;
+    }
+    return p11prov_mlkem_init(vctx, EVP_PKEY_OP_DECAPSULATE, key, params);
+}
+
+static CK_RV p11prov_mlkem_get_lengths(P11PROV_OBJ *key, size_t *ct_len,
+                                       size_t *ss_len)
+{
+    CK_ULONG key_size;
+
+    key_size = p11prov_obj_get_key_size(key);
+
+    *ss_len = 32;
```

Comment:
should we set this in the end of the function, rather than at the beginning to make sure we check the size matches?

Reply:
not sure it make an important difference, but I'll do.